### PR TITLE
テストの実装

### PR DESCRIPTION
**Pull request by tomo42695**
# WHAT
groupモデル、messageモデル、messageコントローラのテスト

# WHY
各モデル及びコントローラの動作確認

# 実装手順
1. 必要gemの導入
2. group_spec.rb, message_spec.rb, message_controller_spec.rbの作成
3. factory girlとfakerの導入
[https://gyazo.com/da7ce0f59ce0ce8b5cdb83bd2e7a020c](https://gyazo.com/da7ce0f59ce0ce8b5cdb83bd2e7a020c)

### CODE REVIEW

File: app/models/group.rb [hunk 1]
```diff
@@ -2,4 +2,5 @@ class Group < ApplicationRecord
   has_many :group_users
   has_many :users, through: :group_users
   has_many :messages
+  validates :name, presence: true
```
Comment (yogoken) [inline app/models/group.rb:6]: 上一行あけといてもいいかも！

File: spec/controllers/messages_controller_spec.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+RSpec.describe MessagesController, type: :controller do
+  let(:group) { create(:group) }
+  describe 'with user logged in' do
+    before do
+      login_user
+    end
+    describe 'GET #new' do
+      let(:request) { get :new, group_id: group }
+      it 'renders the :new template' do
+        request
```
Comment (yogoken) [inline spec/controllers/messages_controller_spec.rb:12]: このrequestって送る必要あるかな？

File: spec/controllers/messages_controller_spec.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+      describe 'with valid params' do
+        let(:request) { post :create, group_id: group, message: { text: 'test' } }
+        it 'redirects to the :new template' do
+          request
+          expect(response).to redirect_to new_group_message_path
+        end
+        it 'should increment the Message count' do
+          expect do
+            request
+          end.to change(Message, :count).by(1)
```
Comment (yogoken) [inline spec/controllers/messages_controller_spec.rb:40]: end.to change????
っていうのはこれちゃんと動いてるのかな？

File: spec/factories/groups.rb [hunk 4]
```diff
@@ -0,0 +1,7 @@
+FactoryGirl.define do
+
```
Comment (yogoken) [inline spec/factories/groups.rb]: いらない改行は消しておこう!!

File: spec/factories/messages.rb [hunk 5]
```diff
@@ -0,0 +1,10 @@
+FactoryGirl.define do
+
```
Comment (yogoken) [inline spec/factories/messages.rb]: ここもだね

File: spec/factories/users.rb [hunk 6]
```diff
@@ -1,8 +1,8 @@
 FactoryGirl.define do
```
Comment (yogoken) [inline spec/factories/users.rb]: ここも！

File: spec/models/group_spec.rb [hunk 7]
```diff
@@ -0,0 +1,14 @@
+require 'rails_helper'
+describe Group do
+  describe '#create' do
+    it "is valid with a name" do
+      group = build(:group)
```
Comment (yogoken) [inline spec/models/group_spec.rb]: group = build(:group)はbefore { group = build(:group) }
にするのはどう？

File: spec/controllers/messages_controller_spec.rb [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+        it 'should show flash alert message' do
+          bad_request
+          expect(flash[:alert]).not_to be_empty
+        end
+      end
+    end
+  end
+
+  describe 'with user logged out' do
+    let(:get_new) { get :new, params: {group_id: group} }
```
Comment (yogoken) [inline spec/controllers/messages_controller_spec.rb]: params: { group_id: group }だね

**Review by yogoken — COMMENTED** [review]
直せればOK
!!

**Comment by yogoken** [comment]
ちなみに rspec -fd で全部のスクショみてみたいな!!

File: app/models/group.rb (hunk 1 above)
Comment (tomo42695) [reply app/models/group.rb:6]: done

File: spec/controllers/messages_controller_spec.rb (hunk 8 above)
Comment (tomo42695) [reply spec/controllers/messages_controller_spec.rb]: done

File: spec/factories/messages.rb (hunk 5 above)
Comment (tomo42695) [reply spec/factories/messages.rb]: done

File: spec/factories/users.rb (hunk 6 above)
Comment (tomo42695) [reply spec/factories/users.rb]: done

File: spec/controllers/messages_controller_spec.rb (hunk 3 above)
Comment (tomo42695) [inline spec/controllers/messages_controller_spec.rb]: 多分…

File: spec/factories/groups.rb (hunk 4 above)
Comment (tomo42695) [reply spec/factories/groups.rb]: done

**Comment by tomo42695** [comment]
DEPRECATION WARNING:           [Devise] including `Devise::TestHelpers` is deprecated and will be removed from Devise.
          For controller tests, please include `Devise::Test::ControllerHelpers` instead.
 (called from <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:3)

MessagesController
  with user logged in
    GET #new
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:10)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:10)
      renders the :new template
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:10)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:10)
      assigns the requested contact to @group
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:10)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:10)
      populates an array of tweets ordered by created_at DESC
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:10)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:10)
      assigns a new message to @message
    POST #create
      with valid params
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:32)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:32)
        redirects to the :new template
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:32)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:32)
        should increment the Message count
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:32)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:32)
        should show flash notice message
      with invalid params
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:49)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:49)
        should not increment the Message count
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:49)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (5 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:49)
        should show flash alert message
  with user logged out
    GET #new
      should not render the :new template
    POST #create
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:72)
DEPRECATION WARNING: ActionController::TestCase HTTP request methods will accept only
keyword arguments in future Rails versions.

Examples:

get :show, params: { id: 1 }, session: { user_id: 1 }
process :update, method: :post, params: { id: 1 }
 (called from block (4 levels) in <top (required)> at /Users/Tomomu_Iwai/projects/chat-space/spec/controllers/messages_controller_spec.rb:72)
      should not increment the Message count

Group
  #create
    is valid with a name
    is invalid without a without a name

Message
  #create
    is valid with a text, image, user_id, group_id
    is invalid without a text

User
  #create
    is valid with a name, email, password, password_confirmation
    is invalid without a name
    is invalid without an email
    is invalid without a password
    is invalid without a password_confirmation although with a password
    is invalid with a duplicate email address

Finished in 0.87139 seconds (files took 3.51 seconds to load)
21 examples, 0 failures

File: spec/controllers/messages_controller_spec.rb (hunk 2 above)
Comment (tomo42695) [inline spec/controllers/messages_controller_spec.rb]: これ書かないとエラーが出ます…

Comment (tomo42695) [reply spec/controllers/messages_controller_spec.rb:12]: 簡略化のために書きました。

File: spec/models/group_spec.rb (hunk 7 above)
Comment (tomo42695) [reply spec/models/group_spec.rb]: これだとエラー出てしまいました。

File: spec/controllers/messages_controller_spec.rb (hunk 3 above)
Comment (tomo42695) [reply spec/controllers/messages_controller_spec.rb:40]: [https://gyazo.com/131ce9a26511d2cc0b05b3f85f665318](https://gyazo.com/131ce9a26511d2cc0b05b3f85f665318)

Comment (tomo42695) [reply spec/controllers/messages_controller_spec.rb:40]: byの引数の数を変えたらエラーが出たので出来ていると思われます！

File: spec/models/group_spec.rb (hunk 7 above)
Comment (tomo42695) [reply spec/models/group_spec.rb]: done

File: spec/controllers/messages_controller_spec.rb (hunk 2 above)
Comment (tomo42695) [reply spec/controllers/messages_controller_spec.rb:12]: beforeで書くと、'populates an array of tweets ordered by created_at DESC'でエラー出ます…

File: spec/models/group_spec.rb [hunk 9]
```diff
@@ -1,15 +1,13 @@
 require 'rails_helper'
+
 describe Group do
+  let (:group) { create(:group)}
```
Comment (yogoken) [inline spec/models/group_spec.rb]: let(:group) { create(:group) }だね

File: spec/models/group_spec.rb [hunk 10]
```diff
@@ -1,15 +1,13 @@
...
-    before do
-      group = build(:group)
-    end
     it "is valid with a name" do
-      group = build(:group)
       expect(group).to be_valid
     end
     it "is invalid without a without a name" do
-      group = build(:group, name: "")
+      group[:name] = ""
```
Comment (yogoken) [inline spec/models/group_spec.rb:10]: おk！！

**Review by yogoken — COMMENTED** [review]
LGTMかな!!

**Pull request merged**